src/test/test-parse-helpers.c

```diff
@@ -92,4 +94,191 @@ TEST(invalid_items) {
+
+        assert_se(r == 0);
+        assert_se(streq(of->path, "/proc/1/ns/mnt"));
+        assert_se(streq(of->fdname, "mnt"));
+        assert_se(of->flags == OPENFILE_RDONLY);
+
+        of = open_file_free(of);
+        r = open_file_parse("../file.dat:file:ro", &of);
+
+        assert_se(r == -EINVAL);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:rw", &of);
+
+        assert_se(r == -EINVAL);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:append", &of);
+
+        assert_se(r == 0);
+        assert_se(streq(of->path, "/proc/1/ns/mnt"));
+        assert_se(streq(of->fdname, "host-mount-namespace"));
+        assert_se(of->flags == OPENFILE_APPEND);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:truncate", &of);
+
+        assert_se(r == 0);
+        assert_se(streq(of->path, "/proc/1/ns/mnt"));
+        assert_se(streq(of->fdname, "host-mount-namespace"));
+        assert_se(of->flags == OPENFILE_TRUNC);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:ro,append", &of);
+
+        assert_se(r == -EINVAL);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:ro,truncate", &of);
+
+        assert_se(r == -EINVAL);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:append,truncate", &of);
+
+        assert_se(r == -EINVAL);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:ignore", &of);
+
+        assert_se(r == 0);
+        assert_se(streq(of->path, "/proc/1/ns/mnt"));
+        assert_se(streq(of->fdname, "host-mount-namespace"));
+        assert_se(of->flags == OPENFILE_IGNORE);
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:ro,ignore", &of);
+
+        assert_se(r == 0);
+        assert_se(streq(of->path, "/proc/1/ns/mnt"));
+        assert_se(streq(of->fdname, "host-mount-namespace"));
+        assert_se(of->flags == (OPENFILE_RDONLY | OPENFILE_IGNORE));
+
+        of = open_file_free(of);
+        r = open_file_parse("/proc/1/ns/mnt:host-mount-namespace:ro:other", &of);
+
+        assert_se(r == -EINVAL);
+}
+
+TEST(open_file_to_string) {
+        _cleanup_free_ char *s = NULL;
+        _cleanup_(open_file_freep) OpenFile *of = NULL;
+
+        of = new0(OpenFile, 1);
+        of->path = strdup("/proc/1/ns/mnt");
+        of->fdname = strdup("host-mount-namespace");
+        of->flags = OPENFILE_RDONLY;
+
+        s = open_file_to_string(of);
+
+        assert_se(streq(s, "/proc/1/ns/mnt:host-mount-namespace:ro"));
+
+        s = mfree(s);
+        of = open_file_free(of);
+        of = new0(OpenFile, 1);
+        of->path = strdup("/proc/1/ns/mnt");
+        of->fdname = strdup("host-mount-namespace");
```

Comment:
Instead of freeing and allocating the entire struct every time, it's sufficient to just replace the fields that need replacing, so something like:

```
assert_se(free_and_strdup(of->path, "/proc/1/ns/mnt"));
...
```